mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+# See README.rst file on addon root folder for license details
+
+from openerp import models, fields, api, _
+from openerp.exceptions import ValidationError
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    @api.one
+    @api.onchange('partner_id')
+    def _onchange_partner(self):
+        if self.partner_id:
+            self.name = self.partner_id.name
+            self.email = self.partner_id.email
+
+    @api.one
+    @api.constrains('partner_id')
+    def _check_partner_id(self):
```

Comment:
This constrains can't be done with SQL because a list can be configured with no mandatory partner, so partner_id could be False in several records.